tensorboard/db.py

```diff
@@ -387,6 +411,333 @@ def _cursor(self):
     return contextlib.closing(self._db_conn.cursor())  # type: sqlite3.Cursor
 
 
+def _is_transient_sqlite_error(exception):
+  """Returns True if transaction should be retried on SQLite.
+
+  :type exception: Exception
+  """
+  return (isinstance(exception, (FakeTransientDatabaseError,
+                                 sqlite3.DatabaseError)) and
+          not isinstance(exception, sqlite3.ProgrammingError))
+
+
+class _TransactionRunner(object):
+  """Utility class for running a transaction attempt.
+
+  If TESTING_MODE is True, then the first invocation will always result
+  in a roll back.
+  """
+
+  def __init__(self, connection, callback):
+    """Runs a write-deferred database transaction.
+
+    :type connection: sqlite3.Connection
+    :type callback: (TransactionConnection) -> T
+    """
+    self._connection = connection
+    self._callback = callback
+    self._should_rollback = TESTING_MODE
+
+  def __call__(self):
+    """Runs a write-deferred database transaction.
+
+    :rtype: T
+    """
+    with self._connection:
+      write_cursor = _DeferredCursor()
+      result = self._callback(
+          TransactionConnection(self._connection, write_cursor))
+      if self._should_rollback:
+        self._should_rollback = False
+        raise FakeTransientDatabaseError()
+      write_cursor.replay(self._connection)
+      return result
+
+
+class FakeTransientDatabaseError(Exception):
+  """Exception thrown to roll back transactions in TESTING_MODE."""
+
+  def __str__(self):
+    return 'FakeTransientDatabaseError'
+
+
+class TransactionConnection(object):
+  """Delegate for PEP 249 Connection object when in a Transaction."""
+
+  def __init__(self, actual_connection, write_cursor):
+    """Creates new instance.
+
+    :type actual_connection: sqlite3.Connection
```

Comment:
Is this required to be a `sqlite3.Connection` or can it be an arbitrary PEP 249 `Connection` object?